taichi/transforms/frontend_type_check.cpp

```diff
@@ -0,0 +1,110 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/frontend_ir.h"
+#include "taichi/ir/statements.h"
+
+namespace taichi::lang {
+
+class FrontendTypeCheck : public IRVisitor {
+ public:
+  explicit FrontendTypeCheck() {
+    allow_undefined_visitor = true;
+  }
+
+  void visit(Block *block) override {
+    std::vector<Stmt *> stmts;
+    // Make a copy since type casts may be inserted for type promotion.
+    for (auto &stmt : block->statements) {
+      stmts.push_back(stmt.get());
+    }
+    for (auto stmt : stmts)
+      stmt->accept(this);
+  }
+
+  void visit(FrontendExprStmt *stmt) override {
+    // Noop
+  }
+
+  void visit(FrontendAllocaStmt *stmt) override {
+    // Noop
+  }
+
+  void visit(FrontendSNodeOpStmt *stmt) override {
+    // Noop
+  }
+
+  void visit(FrontendAssertStmt *stmt) override {
+    if (not stmt->cond->ret_type->is_primitive(PrimitiveTypeID::i32))
+      throw TaichiTypeError(fmt::format(
+          "`assert` conditions must be of type int32; found {}. Consider using "
```

Comment:
The error reporting parts in `FrontendAssertStmt`, `FrontendIfStmt`, `FrontendWhileStmt` seem a bit duplicate. What about abstracting them out (a naive way might be using macros)?

Reply:
Hmm, I'd suggest to extract a function for this

Reply:
Agreed :-)